commit 7a33b226fe0dc243cbed47fcbe49025ec4aea125
Author: Zoltan Herczeg <zherczeg7@gmail.com>
Date:   Sun Jan 25 06:38:11 2026 +0100

    Fix repeat reset for scan prefix (#875)

diff --git a/src/pcre2_jit_compile.c b/src/pcre2_jit_compile.c
index 4cca50d4..5d8c1c14 100644
--- a/src/pcre2_jit_compile.c
+++ b/src/pcre2_jit_compile.c
@@ -6249,6 +6249,7 @@ while (TRUE)
       repeat = GET2(cc, 1);
       if (repeat <= 0)
         {
+        repeat = 1;
         chars_end = chars;
         continue;
         }
diff --git a/src/pcre2_jit_test.c b/src/pcre2_jit_test.c
index 1fc44db5..03cc348b 100644
--- a/src/pcre2_jit_test.c
+++ b/src/pcre2_jit_test.c
@@ -281,6 +281,7 @@ static struct regression_test_case regression_test_cases[] = {
 	{ CM, A, 0, 0, "a1277|a1377|bx487", "bX487" },
 	{ 0, A, 0, 0, "(a|)b*+a", "a" },
 	{ 0, A, 0, 0 | F_NOMATCH, "(.|.|.|.|.)(|.|.|.|.)(.||.|.|.)(.|.||.|.)(.|.|.||.)(.|.|.|.|)(A|.|.|.|.)(.|A|.|.|.)(.|.|A|.|.)(.|.|.|A|.)(.|.|.|.|A)(B|.|.|.|.)(.|B|.|.|.)(.|.|B|.|.)(.|.|.|B|.)(.|.|.|.|B)xa", "1234567890123456ax" },
+	{ 0, A, 0, 0, "(CHAN|LINE)[ab]{0,2}", "LINE" },
 
 	/* Greedy and non-greedy ? operators. */
 	{ MU, A, 0, 0, "(?:a)?a", "laab" },
