diff --git a/execute.c b/execute.c
index 15a487b..489abf5 100644
--- a/execute.c
+++ b/execute.c
@@ -48,6 +48,13 @@
 
 #endif
 
+/* environ is the exported symbol referencing the internal
+   __cygwin_environ variable on cygwin64:
+   <https://cygwin.com/ml/cygwin/2013-06/msg00228.html>.  */
+#if defined __CYGWIN__ && defined __x86_64__
+extern DLL_VARIABLE char **environ;
+#endif
+
 
 #if defined EINTR && ((defined _WIN32 || defined __WIN32__) && ! defined __CYGWIN__)
 
