*** lib/fnmatch.c.bak	2006-07-11 13:54:17.000000000 +0200
--- lib/fnmatch.c	2006-07-30 20:23:09.000000000 +0200
***************
*** 276,282 ****
  fnmatch (const char *pattern, const char *string, int flags)
  {
  # if HANDLE_MULTIBYTE
! #  define ALLOCA_LIMIT 2000
    if (__builtin_expect (MB_CUR_MAX, 1) != 1)
      {
        mbstate_t ps;
--- 276,286 ----
  fnmatch (const char *pattern, const char *string, int flags)
  {
  # if HANDLE_MULTIBYTE
! #  if HAVE_ALLOCA || defined _LIBC
! #   define ALLOCA_LIMIT 2000
! #  else
! #   define ALLOCA_LIMIT 0
! #  endif
    if (__builtin_expect (MB_CUR_MAX, 1) != 1)
      {
        mbstate_t ps;
